### PR TITLE
#1193 Block User

### PR DESCRIPTION
**Pull request by LeraKornienko**

### CODE REVIEW

File: src/app/shared/store/user.actions.ts [hunk 1]
```diff
@@ -279,3 +281,34 @@ export class ResetProviderWorkshopDetails {
...
+  constructor(public payload: BlockedParent) { }
+}
+
+export class BlockParentFail {
+  static readonly type = '[user] block parent fail';
+  constructor(public payload: Error) { }
+}
+export class BlockParentSuccess {
+  static readonly type = '[user] block parent success';
+  constructor(public payload) { }
```
Comment (litvinets) [inline src/app/shared/store/user.actions.ts:295]: public payload: BlockedParent

File: src/app/shared/store/user.state.ts [hunk 2]
```diff
@@ -83,6 +83,14 @@ import {
...
   OnGetProviderByIdFail,
   ResetProviderWorkshopDetails,
+  BlockParent,
+  BlockParentFail,
+  BlockParentSuccess,
+  GetBlockedParents,
+  UnBlockParent,
+  UnBlockParentFail,
+  UnBlockParentSuccess,
+
```
Comment (litvinets) [inline src/app/shared/store/user.state.ts:93]: remove redundant line

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.html [hunk 3]
```diff
@@ -54,8 +54,33 @@ <h4 class="name title">{{ application.child.lastName }} {{ application.child.fir
...
+    <mat-menu #actions="matMenu" class="header_menu">
+      <button *ngIf="!application.isBlocked" mat-menu-item (click)="onBlock(user)">
+        Заблокувати користувача
+      </button>
+      <button *ngIf="application.isBlocked" mat-menu-item (click)="onUnBlock(user)">
+        Розблокувати користувача
+      </button>
+
+
+
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.html:74]: remove redundant lines

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 4]
```diff
@@ -118,6 +138,41 @@ export class ApplicationCardComponent implements OnInit {
...
+   * @param Application application
+   */
+  onBlock(blockedParent: BlockedParent): void {
+    const dialogRef = this.matDialog.open(BlockModalWindowComponent, {});
+    dialogRef.afterClosed().subscribe((result: string)  => {
+      if(result) {
+        const provider: Provider = this.store.selectSnapshot<Provider>(RegistrationState.provider);
+        const parent = this.application.parentId;
+        const reason = result;
+        blockedParent = new BlockedParent( parent, provider.id, reason);
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:152]: let's pass it directly to constructor, so we can remove additional value reason 
  blockedParent = new BlockedParent( parent, provider.id, result);

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 5]
```diff
@@ -118,6 +138,41 @@ export class ApplicationCardComponent implements OnInit {
...
+    /**
+   * This method emit reject Application
+   * @param Application application
+   */
+  onBlock(blockedParent: BlockedParent): void {
+    const dialogRef = this.matDialog.open(BlockModalWindowComponent, {});
+    dialogRef.afterClosed().subscribe((result: string)  => {
+      if(result) {
+        const provider: Provider = this.store.selectSnapshot<Provider>(RegistrationState.provider);
+        const parent = this.application.parentId;
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:150]: const parentId

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts (hunk 4 above)
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:152]: parentId

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 6]
```diff
@@ -118,6 +138,41 @@ export class ApplicationCardComponent implements OnInit {
...
+        const reason = result;
+        blockedParent = new BlockedParent( parent, provider.id, reason);
+        console.log("reason2", reason)
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:153]: remove?

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 7]
```diff
@@ -118,6 +138,41 @@ export class ApplicationCardComponent implements OnInit {
...
+        console.log("blockedParent",blockedParent)
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:154]: remove?

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 8]
```diff
@@ -118,6 +138,41 @@ export class ApplicationCardComponent implements OnInit {
...
+    const dialogRef = this.matDialog.open(ConfirmationModalWindowComponent, {
+      width: Constants.MODAL_SMALL,
+      data: {
+        type: ModalConfirmationType.unBlockParent,
+      }
+    });
+    dialogRef.afterClosed().subscribe((result: string)  => {
+      if(result) {
+        const provider: Provider = this.store.selectSnapshot<Provider>(RegistrationState.provider);
+        const parent = this.application.parentId;
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:170]: parentId

**Review by litvinets — COMMENTED** [review]
overall, there is an excellent job!! Well done, comments are related only to code style

**Review by Pabloostr — APPROVED** [review]

**Review by sivxkxn — CHANGES_REQUESTED** [review]
There are some minor comments, in general, the PR is good ;)

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 9]
```diff
@@ -115,6 +133,38 @@ export class ApplicationCardComponent implements OnInit {
     });
   }
 
+    /**
+   * This method emit reject Application
+   * @param Application application
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:138]: fix js doc according to the method

File: src/app/shell/personal-cabinet/cabinet-data/cabinet-data.component.ts [hunk 10]
```diff
@@ -53,9 +53,9 @@ export abstract class CabinetDataComponent implements OnInit, OnDestroy {
   filteredChildren: Child[]
 
   constructor(
-    public store: Store, 
-    public matDialog: MatDialog, 
-    protected actions$: Actions, 
+    public store: Store,
+    public matDialog: MatDialog,
+    protected actions$: Actions,
```
[the remark this answers is not in the document]
Comment (litvinets) [reply src/app/shell/personal-cabinet/cabinet-data/cabinet-data.component.ts:58]: we are following the approach from documentation, there is an actual observable
image
image

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.html [hunk 11]
```diff
@@ -36,7 +36,7 @@ <h4 class="name title">{{ application.child.lastName }} {{ application.child.fir
         [ngClass]="trigger.menuOpen ? 'activeInfoBtn' : 'inactiveInfoBtn'">info_outline</i></span>
         <mat-menu class="status-menu"  #menu="matMenu" disabled [overlapTrigger]="false"
           xPosition="after" class="mat-application-menu">
-          <app-status-info-card [application]="application" [status]="status"></app-status-info-card>
+          <!-- <app-status-info-card [application]="application" [blockedParent]="blockedParent" [status]="status"></app-status-info-card> -->
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.html:39]: remove?

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.html [hunk 12]
```diff
@@ -54,8 +54,33 @@ <h4 class="name title">{{ application.child.lastName }} {{ application.child.fir
...
+
+    </mat-menu>
+    <mat-menu #email="matMenu" class="header_menu">
+      <button mat-menu-item (click)="sendInvitation()">
+        Надіслати запрошення
+      </button>
+    </mat-menu>
   </mat-card-content>
 
+
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.html:83]: redundant line

**Review by litvinets — CHANGES_REQUESTED** [review]

File: src/app/shared/components/block-modal-window/block-modal-window.component.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  readonly validationConstants= ValidationConstants;
+
+  @Input() application: Application;
+
+  ReasonFormControl= new FormControl('', [
+    Validators.required,
+    Validators.minLength(ValidationConstants.INPUT_LENGTH_1),
+    Validators.minLength(ValidationConstants.MAX_DESCRIPTION_LENGTH_500)
+  ]);
+  modalTitle = 'ЗАБЛОКУВАТИ';
```
[the remark this answers is not in the document]
Comment (litvinets) [reply src/app/shared/components/block-modal-window/block-modal-window.component.ts:29]: create related task https://github.com/ita-social-projects/OoS-Frontend/issues/1447

**Review by sivxkxn — APPROVED** [review]

**Pull request merged**